common/login.js

```diff
@@ -6,68 +6,214 @@
+                                var q = [menuConfig];
+                                while (q.length > 0) {
+                                    var option = q.shift();
+
+                                    var parentNewTab = option.newTab;
+                                    var parentAcls = option.acls;
+
+                                    // template the url
+                                    // NOTE: Like in navbar.js, this is done here to prevent writing a recursive function (again) in `setConfigJSON()`
+                                    if (option.urlPattern && isValueDefined(catalogId)) {
+                                        option.url = ERMrest.renderHandlebarsTemplate(option.urlPattern, null, {id: catalogId});
+
+                                        // only append pcid/ppid if link is to a chaise url
+                                        if (MenuUtils.isChaise(option.url, ConfigUtils.getContextJSON())) {
+                                            option.url = MenuUtils.addLogParams(option.url, ConfigUtils.getContextHeaderParams());
+                                        }
+                                    }
+
+                                    // If current node has children, set each child's newTab to its own existing newTab or parent's newTab
+                                    // used to set ACLs for each child as well
+                                    if (Array.isArray(option.menuOptions) || Array.isArray(option.children)) {
+                                        var arr = option.menuOptions || option.children;
+                                        arr.forEach(function (child) {
+                                            // get newTab from the parent
+                                            if (child.newTab === undefined) child.newTab = parentNewTab;
+                                            // if we have to open in newtab
+                                            if (forceNewTab) child.newTab = true;
+
+                                            // get acls settings from the parent
+                                            if (child.acls === undefined) {
+                                                child.acls = parentAcls;
+                                            } else {
+                                                // acls could be defined with nothing in it, or with only show or only enable
+                                                if (child.acls.show === undefined) child.acls.show = parentAcls.show;
+                                                if (child.acls.enable === undefined) child.acls.enable = parentAcls.enable;
+                                            }
+
+                                            q.push(child);
+                                        });
+                                    }
+                                    option.isValid = MenuUtils.isOptionValid(option);
```

Comment:
We're going to have to rewrite this in React so you don't need to change this, but I feel like both code and template would have been simpler/more efficient if you were just removing invalid elements. Instead of just setting a boolean that then template removes.

I think we're doing the same thing in navbar.

Reply:
In navbar, we are not removing the option. We are using ng-if to decide if certain properties are set and show that `<li>` if the condition is met. 

The way we are using a queue to modify each menu option, we only have access to a reference to the menuOption from the array or from the children array. Because of this, we can't simply "delete" that entry if it is invalid, so I decided on a similar approach which is to mark the item as valid/invalid and hide it if it is invalid.